libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp

```diff
@@ -0,0 +1,156 @@
+         if( indx >= _end_iterator_to_table.size() ) return nullptr;
+         return &_end_iterator_to_table[indx];
+      }
+
+      const secondary_obj_type& get( int iterator ) const {
+         validate_object_iterator(iterator, "dereference of end iterator");
+         // grab a const ref, to ensure that we are returning a reference to the actual object in the vector
+         const auto& result = _iterator_to_object[iterator];
+         EOS_ASSERT( result, table_operation_not_permitted, "dereference of deleted object" );
+         return *result;
+      }
+
+      void remove( int iterator ) {
+         validate_object_iterator(iterator, "cannot call remove on end iterators");
+
+         auto optional_obj = _iterator_to_object[iterator];
+         if( !optional_obj ) return;
+         _object_to_iterator.erase( *optional_obj );
+
+         _iterator_to_object[iterator].reset();
+      }
+
+      void swap( int iterator, const secondary_key_type& secondary, uint64_t payer ) {
+         validate_object_iterator(iterator, "cannot call swap on end iterators");
+
+         auto& optional_obj = _iterator_to_object[iterator];
+         if( !optional_obj ) return;
+         // check to see if the object will actually change
+         if(optional_obj->payer == payer && optional_obj->secondary == secondary) return;
+         const bool map_key_change = optional_obj->secondary != secondary;
+         if (map_key_change) {
+            // edit object and swap out the map entry
+            _object_to_iterator.erase( *optional_obj );
+            optional_obj->secondary = secondary;
+         }
+         optional_obj->payer = payer;
+
+         if (map_key_change) {
+            _object_to_iterator.insert({ *optional_obj, iterator });
+         }
+      }
+
+      int add( const secondary_obj_type& obj ) {
+         auto itr = _object_to_iterator.find( obj );
+         if( itr != _object_to_iterator.end() )
+              return itr->second;
+
+         EOS_ASSERT( obj.table_ei < _invalid_iterator, invalid_table_iterator, "not an end iterator" );
+         const auto indx = end_iterator_to_index(obj.table_ei);
+         EOS_ASSERT( indx < _end_iterator_to_table.size(), invalid_table_iterator, "an invariant was broken, table should be in cache" );
+         _object_to_iterator.insert({ obj, _iterator_to_object.size() });
+         _iterator_to_object.emplace_back( obj );
+
+         return _iterator_to_object.size() - 1;
+      }
+
+   private:
+      void validate_object_iterator(int iterator, const char* explanation_for_no_end_iterators) const {
+         EOS_ASSERT( iterator != _invalid_iterator, invalid_table_iterator, "invalid iterator" );
+         EOS_ASSERT( iterator >= 0, table_operation_not_permitted, explanation_for_no_end_iterators );
+         EOS_ASSERT( (size_t)iterator < _iterator_to_object.size(), invalid_table_iterator, "iterator out of range" );
+      }
+      unordered_map<unique_table, int, unique_table_hash > _table_cache;
```

Comment:
`unordered_map` is vulnerable to DOS attacks caused by hash collisions.

Reply:
changing to map